### PR TITLE
[otap-dataflow] updated kafka receiver metrics and logging

### PR DESCRIPTION
**Pull request by c1ly**
Updated Kafka receiver to track the following metrics

 - rebalances_total (counter)
 - partitions_assigned (gauge)
 - consumer_lag  (gauge) — (offset delta between high watermark and committed offset per partition

Renamed:
 - partitions_assigned (counter) -> partition_assignments
 - partitions_revoked -> partitions_revocations

Added log events when:
 - Partition assignment occurs, emits list of partition IDs assigned
 - Partition revocation occurs emits list of partition IDs revoked
 - Consumer joins or leaves consumer group

### CODE REVIEW

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs [hunk 1]
```diff
@@ -967,6 +1062,21 @@ impl KafkaReceiver {
...
+                _ = async {
+                    match lag_ticker.as_mut() {
+                        Some(ticker) => ticker.tick().await,
+                        // Unreachable: the branch guard keeps this future from
+                        // being polled when the ticker is disabled.
+                        None => std::future::pending().await,
+                    }
+                }, if lag_ticker.is_some() && draining_deadline.is_none() => {
+                    self.refresh_consumer_lag(&consumer, LAG_FETCH_TIMEOUT);
+                }
```
[the remark this answers is not in the document]
Comment (c1ly) [reply rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs]: will follow up, consumer lag is opt in to prevent blocking the kafka recv() when not wanted

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/config.rs [hunk 2]
```diff
@@ -517,6 +532,10 @@ impl TryFrom<KafkaReceiverConfigBuilder> for KafkaReceiverConfig {
             return Err("commit.interval_ms, when set, must be > 0".to_string());
         }
 
+        if builder.lag_refresh_interval_ms == Some(0) {
+            return Err("lag_refresh_interval_ms, when set, must be > 0".to_string());
```
Comment (lquerel) [inline rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/config.rs]: In general in the project, we prefer to use structured errors (thiserror).

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/config.rs [hunk 3]
```diff
@@ -275,6 +275,21 @@ pub struct KafkaReceiverConfigBuilder {
...
+    /// When set (and manual commit is active), the receiver periodically queries
+    /// the broker high watermark for each owned partition and reports the average
+    /// per-partition lag via the `consumer_lag` gauge. The refresh runs on a
+    /// dedicated periodic-timer branch of the receive loop; each per-partition
+    /// high-watermark lookup is a bounded blocking broker call. Because the
+    /// lookups run inline on the receive loop, a refresh can briefly pause
+    /// message processing under partition fanout or broker slowness, so this is
+    /// intended for infrequent (opt-in) intervals.
+    ///
+    /// Defaults to `None` (consumer-lag refresh disabled).
```
Comment (lquerel) [inline rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/config.rs]: When enabled, what is the recommended consumer-lag refresh duration? Please add the answer to the comment and documentation.

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs [hunk 4]
```diff
@@ -967,6 +1065,21 @@ impl KafkaReceiver {
...
+                _ = async {
+                    match lag_ticker.as_mut() {
+                        Some(ticker) => ticker.tick().await,
+                        // Unreachable: the branch guard keeps this future from
+                        // being polled when the ticker is disabled.
+                        None => std::future::pending().await,
+                    }
+                }, if lag_ticker.is_some() && draining_deadline.is_none() => {
+                    self.refresh_consumer_lag(&consumer, LAG_FETCH_PARTITION_TIMEOUT);
+                }
```
Comment (lquerel) [inline rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs:1113]: This timer branch still runs on the pipeline-local receive loop. refresh_consumer_lag calls the blocking fetch_watermarks operation sequentially for every tracked partition, with a five-second timeout per call. While it runs, this core cannot process Kafka records, ack/nack and control messages, telemetry, or shutdown; N slow partitions can therefore pause it for up to N * 5 seconds. Opt-in configuration limits who encounters the problem but does not preserve runtime responsiveness when enabled. Please perform this as bounded, cancellation-aware work with at most one refresh in flight and a total refresh deadline, returning only the completed sample to the local loop.

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs [hunk 5]
```diff
@@ -531,6 +545,73 @@ impl KafkaReceiver {
...
+        &mut self,
+        consumer: &StreamConsumer<C>,
+        timeout: Duration,
+    ) {
+        if self.config.is_auto_commit() {
+            return;
+        }
+
+        // Committed offsets we would persist for each owned partition. Using the
+        // tracker snapshot avoids an extra broker `committed()` round-trip.
```
Comment (lquerel) [inline rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs]: committable_snapshot() is the offset the local tracker would submit, not the consumer group offset acknowledged by Kafka. Normal commits use CommitMode::Async, so this snapshot advances before the commit callback succeeds and remains advanced when Kafka rejects the commit. For example, with high watermark 1000, local committable offset 1000, and broker-committed offset 100, this reports zero lag instead of 900. Please derive consumer_lag from broker-acknowledged committed offsets, either through a batched query or state updated by successful commit callbacks, or name and document this as local processing lag instead.

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs [hunk 6]
```diff
@@ -531,6 +545,73 @@ impl KafkaReceiver {
...
+                    );
+                }
+            }
+        }
+
+        let avg = if counted == 0 {
+            0.0
+        } else {
+            total_lag as f64 / counted as f64
+        };
```
Comment (lquerel) [inline rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs]: When every watermark lookup fails, counted remains zero and this overwrites the gauge with 0.0, converting an unavailable observation into "no lag". Partial failures similarly calculate an average over only the successful partitions and can understate lag. Please retain the previous sample or otherwise represent missing data without publishing zero; the refresh failure can remain observable through the error metric/event.

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs [hunk 7]
```diff
@@ -502,12 +510,18 @@ impl KafkaReceiver {
 
         let delta = self.rebalance_state.drain_metrics();
         if !delta.is_empty() {
+            self.metrics.rebalances_total.add(delta.rebalances_total);
```
Comment (lquerel) [inline rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs:516]: These new rebalance metrics and log events are never produced with `commit.mode: auto`. `reconcile_rebalance_state` returns here, and both rebalance callbacks also short-circuit before handle_assign/handle_revoke, leaving the counters and ownership gauge at zero and suppressing the join/leave events. The README only identifies consumer_lag as manual-only. Please decouple assignment observability from manual offset bookkeeping, or explicitly scope all these instruments and events to manual mode.

**Review by lquerel — CHANGES_REQUESTED** [review]
I believe there are few issues to fix before merging this PR. Thanks

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/config.rs (hunk 3 above)
Comment (c1ly) [reply rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/config.rs]: set recommended to 60s

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs (hunk 7 above)
Comment (c1ly) [reply rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs:516]: documented as manual mode only

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs (hunk 4 above)
Comment (c1ly) [reply rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs:1113]: performs calculation in spawned task with a deadline to prevent blocking the main receive loop

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs (hunk 6 above)
Comment (c1ly) [reply rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs]: added a counter to track high watermarks from each partition and any failed attempts at retrieving data from the consumer to avoid any bad calculation of consumer_lag -> doesn't update the metric and instead emits a warning

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/config.rs [hunk 8]
```diff
@@ -46,6 +46,25 @@ pub(crate) const MANAGED_CONSUMER_CONFIG_KEYS: &[&str] = &[
     "debug",
 ];
 
+/// Error returned when validating a [`KafkaReceiverConfigBuilder`] into a
+/// [`KafkaReceiverConfig`](super::config::KafkaReceiverConfig).
+#[derive(Debug, thiserror::Error)]
+pub enum KafkaReceiverConfigError {
```
Comment (lquerel) [inline rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/config.rs]: You probably have other errors into the kafka-receiver module and you should combined all of them in a single error enum.

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs [hunk 9]
```diff
@@ -967,6 +1122,67 @@ impl KafkaReceiver {
...
+                // self-reports any failure), so publishing is a single `set`.
+                result = async {
+                    match lag_refresh_in_flight.as_mut() {
+                        Some(handle) => {
+                            tokio::time::timeout(LAG_REFRESH_TOTAL_DEADLINE, handle).await
+                        }
+                        // Unreachable: guarded by `is_some()` below.
+                        None => std::future::pending().await,
+                    }
+                }, if lag_refresh_in_flight.is_some() => {
```
Comment (lquerel) [inline rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs]: This is not currently a reliable total deadline. The surrounding select! is biased (and complicated), and consumer.recv() appears before this branch. While Kafka records remain available, this branch can be starved, so a completed refresh is never published. Also, timeout(...) is reconstructed whenever the outer loop restarts, resetting its 15s clock.

Please store an absolute tokio::time::Instant deadline with the in-flight handle, use timeout_at(deadline, handle)`, and poll this branch before consumer.recv(). Add a paused-time test that keeps the receive branch ready and verifies that completion and timeout are still processed.

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs [hunk 10]
```diff
@@ -967,6 +1122,67 @@ impl KafkaReceiver {
...
+                    lag_refresh_in_flight = None;
```
Comment (lquerel) [inline rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs]: Dropping a JoinHandle does not cancel an already-started `spawn_blocking` task. After this timeout, the state becomes None, allowing a later tick to start another refresh while the old task is still querying partitions. Shutdown has the same problem.

You should keep the refresh marked in flight until the worker actually exits. Pass the absolute deadline into the worker, check it between partitions, and limit each broker call to the remaining time. Do not start another refresh until the previous worker finishes. Add a test proving that at most one worker runs at a time.

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs [hunk 11]
```diff
@@ -531,6 +553,102 @@ impl KafkaReceiver {
...
+                    }
+                };
+
+            // Per-partition consumer-group lag for every owned+committed
+            // partition. Any failed broker read abandons the whole refresh (fail
+            // fast) so a mean is never computed from incomplete data.
+            let mut per_partition_lag: Vec<i64> = Vec::new();
+            for elem in &committed.elements() {
+                // Skip partitions with no broker-committed offset yet
+                // (`Offset::Invalid`); they are not part of the lag measurement.
```
Comment (lquerel) [inline rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs]: Skipping `Offset::Invalid` calculates the mean from only a subset of owned partitions, although the metric contract says it covers all owned partitions. If every offset is invalid (or the assignment becomes empty) the previous gauge value is retained and may become stale.

Please abort the refresh when any owned partition lacks a committed offset and report the refresh as incomplete. Reset `consumer_lag` to the documented empty-assignment value when ownership becomes zero. Add tests for empty, all-invalid, and mixed valid/invalid assignments.

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/rebalance.rs [hunk 12]
```diff
@@ -413,28 +447,44 @@ impl RebalanceState {
...
         }
 
         let _ = self
-            .partitions_revoked
+            .partition_revocations
             .fetch_add(owned_revoked, Ordering::Relaxed);
+
+        // Structured observability: list the genuinely-owned revoked partition
+        // IDs, and emit a consumer-group "left" event when this revocation drops
+        // the consumer to zero owned partitions.
```
Comment (lquerel) [inline rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/rebalance.rs]: `owned_after == 0` means (IIUC) that the consumer currently owns no partitions. It does not mean that it left the consumer group. An eager rebalance revokes all partitions before reassigning them, so an ordinary rebalance produces a misleading left/joined pair.

IMO, you should rename these as assignment state events, such as assignment.became_empty and assignment.became_non_empty, and update the README and changelog. As actual group membership events are probably required, they must come from membership state rather than assignment size.

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/rebalance.rs [hunk 13]
```diff
@@ -405,6 +433,12 @@ impl RebalanceState {
         // the newly-added count in `set_assignment`.
         let mut revoked_tagged = Vec::with_capacity(revoked.len());
         let mut owned_revoked = 0u64;
+        // Log buffer for the genuinely-owned revoked partitions only, built
+        // inline so the logged list matches `owned_revoked` (a revoke reported
+        // for a partition this consumer did not own is neither counted nor
+        // listed).
+        let mut revoked_list = String::new();
```
Comment (lquerel) [inline rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/rebalance.rs]: This string grows with every revoked partition and is built while the assignment mutex is held. Large assignments can therefore allocate and emit very large log events while delaying rebalance bookkeeping.

Please use a shared bounded formatter for both revoked and acquired lists. Cap the entry count or encoded size and include count, listed_count, and truncated fields. Add a test covering the truncation behavior.

**Review by lquerel — COMMENTED** [review]
Added few additional issues

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs (hunk 9 above)
Comment (c1ly) [reply rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs]: updated the consumer_lag calculation to keep track of the deadline, returns none upon deadline exceeded to complete the spawned task. Deadline is checked during each partition calculation. Also separated the tokio select blocks so one block spawns the worker while there are no in flight workers, while the other polls the worker to completion and applies the result

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs (hunk 10 above)
Comment (c1ly) [reply rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs]: worker keeps track of the instant deadline

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs (hunk 11 above)
Comment (c1ly) [reply rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs]: consumer lag calculation returns 0 upon detecting no assignment, keep previous value when offset is invalid

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/rebalance.rs (hunk 12 above)
Comment (c1ly) [reply rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/rebalance.rs]: updated naming

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/rebalance.rs (hunk 13 above)
Comment (c1ly) [reply rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/rebalance.rs]: added PartitionListFmt struct that is used to build the string and caps the entries at 64. Any additional entries update count and sets the truncated status to true. Truncated strings end with a trailing `...` to signify that there additional partitions not displayed

File: rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs [hunk 14]
```diff
@@ -685,13 +757,18 @@ impl KafkaReceiver {
...
                                         "kafka.shutdown.commit_failed",
                                         error = %e,
                                     );
                                 }
                             }
+                            // Drain any in-flight consumer-lag worker so we do not
+                            // abandon a running `spawn_blocking` task
+                            if let Some((handle, lag_deadline)) = lag_refresh_in_flight.take() {
+                                let _ = tokio::time::timeout_at(lag_deadline, handle).await;
+                            }
```
Comment (lquerel) [inline rust/otap-dataflow/crates/contrib-nodes/src/receivers/kafka_receiver/receiver.rs:771]: For a future PR: This wait is bounded by the lag-refresh deadline, not the pipeline shutdown deadline received above. A recently started refresh can therefore delay shutdown past the requested deadline. If lag_deadline has already expired, the handle is dropped while the non-cancellable spawn_blocking task continues, contradicting the drain comment. Bound the wait by min(lag_deadline, shutdown_deadline) and provide cooperative cancellation if the worker must not outlive the receiver. Please add a short-deadline shutdown test.

**Review by lquerel — APPROVED** [review]
I added one comment for a future PR. Please create a GH issue to keep track of it. Thanks
